File: src/services/PluginLoader.ts
import { Plugin, PluginMetadata, AppConfig } from '../types/plugin.types';

export class PluginLoader {
  private static config: AppConfig | null = null;

  /**
   * Initialize the plugin loader with config
   */
  static initialize(config: AppConfig) {
    this.config = config;
  }

  /**
   * Load a plugin by its metadata
   */
  static loadPlugin(metadata: PluginMetadata): Plugin | null {
    if (!this.config) {
      console.error('PluginLoader not initialized. Call PluginLoader.initialize(config) first.');
      return null;
    }

    const component = this.config.componentRegistry[metadata.component];
    
    if (!component) {
      console.warn(`Plugin component not found for component key: ${metadata.component}`);
      return null;
    }

    return {
      metadata,
      component,
    };
  }

  /**
   * Load multiple plugins by their metadata
   */
  static loadPlugins(pluginMetadata: PluginMetadata[]): Plugin[] {
    return pluginMetadata
      .map(metadata => this.loadPlugin(metadata))
      .filter((plugin): plugin is Plugin => plugin !== null);
  }

  /**
   * Get available plugin IDs
   */
  static getAvailablePluginIds(): string[] {
    if (!this.config) {
      return [];
    }
    return this.config.plugins.map(plugin => plugin.id);
  }

  /**
   * Check if a plugin is available
   */
  static isPluginAvailable(pluginId: string): boolean {
    if (!this.config) {
      return false;
    }
    return this.config.plugins.some(plugin => plugin.id === pluginId);
  }
}

File: Dashboard.Blazor/Services/PluginLoader.cs
using Dashboard.Blazor.Models;

namespace Dashboard.Blazor.Services
{
    public class PluginLoader
    {
        private AppConfig _appConfig = new();
        private Dictionary<string, Plugin> _loadedPlugins = new();

        public void Initialize(AppConfig config)
        {
            _appConfig = config;
        }

        public List<Plugin> LoadPlugins(List<PluginMetadata> pluginMetadata)
        {
            var plugins = new List<Plugin>();

            foreach (var metadata in pluginMetadata)
            {
                if (_appConfig.ComponentRegistry.TryGetValue(metadata.Component, out var componentType))
                {
                    var plugin = new Plugin
                    {
                        Metadata = metadata,
                        ComponentType = componentType
                    };

                    plugins.Add(plugin);
                    _loadedPlugins[metadata.Id] = plugin;
                }
            }

            return plugins;
        }

        public Plugin? GetPlugin(string pluginId)
        {
            return _loadedPlugins.TryGetValue(pluginId, out var plugin) ? plugin : null;
        }

        public PluginHierarchy GeneratePluginHierarchy(List<PluginMetadata> plugins)
        {
            var hierarchy = new PluginHierarchy();

            foreach (var plugin in plugins)
            {
                if (!hierarchy.ContainsKey(plugin.Category))
                {
                    hierarchy[plugin.Category] = new List<PluginMetadata>();
                }
                hierarchy[plugin.Category].Add(plugin);
            }

            return hierarchy;
        }
    }
}